src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-            return 0;
-        }
-
-        final Func1<InnerSubscriber<T>, Boolean> DRAIN_ACTION = new Func1<InnerSubscriber<T>, Boolean>() {
-
-            @Override
-            public Boolean call(InnerSubscriber<T> s) {
-                if (s.q != null) {
-                    long r = mergeProducer.requested;
-                    int emitted = s.drainQueue();
-                    if (emitted > 0) {
-                        s.requestMore(emitted);
+                } finally {
+                    if (!skipFinal) {
+                        synchronized (this) {
+                            emitting = false;
+                        }
                     }
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
+            } else {
+                // either someone else is emitting or there was no available request
+                // queue up the scalar
+                RxRingBuffer svq = scalarQueue;
+                if (svq == null) {
+                    svq = RxRingBuffer.getSpscInstance();
+                    add(svq);
+                    scalarQueue = svq;
+                }
+                try {
+                    svq.onNext(nl.next(value));
+                } catch (MissingBackpressureException ex) {
+                    onError(ex);
+                    return;
+                } catch (IllegalStateException ex) {
+                    // if there was an IAE and we are unsubscribed, just quit
+                    if (!isUnsubscribed()) {
+                        onError(ex);
                     }
+                    return;
                 }
-                return Boolean.TRUE;
-            }
-
-        };
-
-        @Override
-        public void onError(Throwable e) {
-            if (!completed) {
-                completed = true;
-                innerError(e, true);
+                // then try to emit it (or notify the running emission loop).
+                emit();
             }
         }
         
-        private void innerError(Throwable e, boolean parent) {
-            if (delayErrors) {
-                synchronized (this) {
-                    if (exceptions == null) {
-                        exceptions = new ConcurrentLinkedQueue<Throwable>();
+        void fastEmitNext(T value) {
+            synchronized (this) {
+                if (emitting) {
+                    int s = fastSize;
+                    Object[] o = fastArray;
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
+                        Object[] o2 = new Object[s + (s >> 2)];
+                        System.arraycopy(o, 0, o2, 0, s);
```

Comment:
I suggest that the solution is to not use a growable array, especially since we shouldn't have an unbounded buffer anyways. This is why we have pooled `RxRingBuffer`.